nox/manifest.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2017 Jon Wayne Parrott
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+import copy
+import itertools
+
+from nox._parametrize import generate_calls
+from nox.logger import logger
+from nox.sessions import Session
+
+
+class Manifest(object):
+    """Session manifest.
+
+    The session manifest provides the final list of sessions that should be
+    run by nox.
+
+    It is possible for this to be mutated during execution. This allows for
+    useful use cases, such as for one session to "notify" another.
+
+    Args:
+        session_functions (Mapping[str, function]): The registry of discovered
+            session functions.
+        global_config (): The global configuration.
+    """
+    def __init__(self, session_functions, global_config):
+        self._all_sessions = []
+        self._config = global_config
+
+        # Create the sessions based on the provided session functions.
+        for name, func in session_functions.items():
+            self.make_session(name, func)
+
+        # By default, every available session starts off in the queue.
+        self._queue = copy.copy(self._all_sessions)
```

Comment:
Basically the [don't make me think](https://en.wikipedia.org/wiki/Don't_Make_Me_Think) principle. The `[::]` idiom, while certainly used, is not something that I think most people see every day, and I still have to do cognitive work to say, "oh right; it is making a shallow copy".

Or I can just import `copy` and explicitly write, "I am making a shallow copy." I believe this is better for readability.